### PR TITLE
[14.0][ADD] purchase_requisition_from_rfq

### PR DESCRIPTION
**Pull request by ps-tubtim**
This module helps to create a requisition from RFQs.

Process:
1. Go to menu Purchase > Orders > Requests for Quotation (RFQ)
2. Select multiple RFQ(s) that want to create purchase agreement for
3. Click Action > Create Purchase Agreement from RFQ
4. Confirm and validate the purchase agreement
5. Click smart button RFQs/Orders and confirm on the winning RFQ
6. Return to Purchase Agreement and Click Close

### CODE REVIEW

**Comment by kittiu** [comment]
> This PR is waiting for #779, you can always use this for readability.
> 
> * [x]  [12.0][ADD] purchase_requisition_from_rfq

I closed v12 of this PR. Let's starts here.

**Comment by kittiu** [comment]
I created RFQ with analytic account, and then created TE. And it seem analytic is not being passed here.
Please take analytic account into consideration too.

**Comment by kittiu** [comment]
@ps-tubtim any update?

**Comment by ps-tubtim** [comment]
@kittiu I updated this module.

File: purchase_requisition_from_rfq/hooks.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2019 Ecosoft Co., Ltd (http://ecosoft.co.th/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo.addons.purchase_requisition.models.purchase import PurchaseOrder
+
+
+def post_load_hook():
```
Comment (newtratip) [inline purchase_requisition_from_rfq/hooks.py]: @ps-tubtim, I saw this function. I didn't sure what is this for. Could you more explain this function pls ?

File: purchase_requisition_from_rfq/README.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
+============================
```
Comment (newtratip) [inline purchase_requisition_from_rfq/README.rst:1]: Please re-genarate readme for 14.0

File: purchase_requisition_from_rfq/views/purchase_requisition_view.xml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+
+    <record model="ir.ui.view" id="view_purchase_requisition_form_inherit">
```
Comment (newtratip) [inline purchase_requisition_from_rfq/views/purchase_requisition_view.xml]:
```suggestion
    <record id="view_purchase_requisition_form" model="ir.ui.view">
```

File: purchase_requisition_from_rfq/views/purchase_requisition_view.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        <field name="name">purchase.requisition.form.inherit</field>
+        <field name="model">purchase.requisition</field>
+        <field
+            name="inherit_id"
+            ref="purchase_requisition.view_purchase_requisition_form"
+        />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='is_quantity_copy']" position="before">
+                <field name="from_rfq" invisible='1' />
```
Comment (newtratip) [inline purchase_requisition_from_rfq/views/purchase_requisition_view.xml]:
```suggestion
                <field name="from_rfq" invisible="1" />
```

File: purchase_requisition_from_rfq/views/purchase_view.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+
+    <record
+        id="action_purchase_purchase_agreement_from_rfq"
```
Comment (newtratip) [inline purchase_requisition_from_rfq/views/purchase_view.xml]: id="action_purchase_agreement_from_rfq" ?

File: purchase_requisition_from_rfq/models/purchase.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    def button_confirm(self):
+        res = super(PurchaseOrder, self).button_confirm()
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase.py]:
```suggestion
        res = super().button_confirm()
```

File: purchase_requisition_from_rfq/models/purchase.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        for order in self:
+            requisition = order.requisition_id
+            if order.requisition_id and requisition.from_rfq:
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase.py]:
```suggestion
            if requisition and requisition.from_rfq:
```

File: purchase_requisition_from_rfq/models/purchase.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                requisition.vendor_id = order.partner_id.id
+                for line in requisition.line_ids:
+                    price_unit = order.order_line.filtered(
+                        lambda l: l.product_id == line.product_id
+                    ).mapped("price_unit")
+                    if price_unit:
+                        line.price_unit = max(price_unit)
+        return res
+
+    def _auto_cancel_another_order(self):
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase.py:25]: self.ensure_one()

File: purchase_requisition_from_rfq/models/purchase.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        requisition_id = self.requisition_id.id
+        purchases = self.env["purchase.order"].search(
+            [("requisition_id", "=", requisition_id)]
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase.py]: requisition_id used only this line, I think should be as below
```suggestion
            [("requisition_id", "=", self.requisition_id.id)]
```

and remove line 26

File: purchase_requisition_from_rfq/models/purchase.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        )
+        for purchase in purchases:
+            if purchase.state != "purchase":
+                purchase.state = "cancel"
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase.py]: We can use short code
`purchases.filtered(lambda l: l.state != "purchase").state = "cancel"`

ps. Please test before you change.

File: purchase_requisition_from_rfq/models/purchase_requisition.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    purchase_ids = fields.One2many(
+        default=lambda self: self._context.get("active_ids", []),
+    )
+    line_ids = fields.One2many(
+        default=lambda self: self._get_line(),
+    )
+    from_rfq = fields.Boolean(
+        default=False,
+        readonly=True,
+        help="True, if this purchase agreement is created from RFQs",
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase_requisition.py:21]: Please add copy=False

File: purchase_requisition_from_rfq/models/purchase_requisition.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    )
+
+    def _get_line(self):
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase_requisition.py:24]: `active_ids` was used in 2 positions, I think should declare variable `active_ids`

active_ids = self._context.get("active_ids", [])

File: purchase_requisition_from_rfq/models/purchase_requisition.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        if self._context.get("active_ids"):
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase_requisition.py]:
```suggestion
        if active_ids:
```

File: purchase_requisition_from_rfq/models/purchase_requisition.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+            orders = self.env["purchase.order"].browse(
+                self._context.get("active_ids", [])
```
Comment (newtratip) [inline purchase_requisition_from_rfq/models/purchase_requisition.py]:
```suggestion
                active_ids
```

**Comment by newtratip** [comment]
@ps-tubtim, Could you check pre-commit and CI ?

**Comment by kittiu** [comment]
@ps-tubtim please attend the comment.

File: purchase_requisition_from_rfq/README.rst (hunk 2 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/README.rst:1]: Fixed.

File: purchase_requisition_from_rfq/views/purchase_requisition_view.xml (hunk 3 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/views/purchase_requisition_view.xml]: Fixed.

File: purchase_requisition_from_rfq/views/purchase_requisition_view.xml (hunk 4 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/views/purchase_requisition_view.xml]: Fixed.

File: purchase_requisition_from_rfq/models/purchase.py (hunk 6 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py]: Fixed.

File: purchase_requisition_from_rfq/models/purchase.py (hunk 7 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py]: Fixed.

File: purchase_requisition_from_rfq/views/purchase_view.xml (hunk 5 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/views/purchase_view.xml]: Fixed.

File: purchase_requisition_from_rfq/models/purchase.py (hunk 8 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py:25]: Fixed.

File: purchase_requisition_from_rfq/models/purchase.py (hunk 9 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py]: Fixed.

File: purchase_requisition_from_rfq/models/purchase.py (hunk 10 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py]: Fixed.

File: purchase_requisition_from_rfq/models/purchase_requisition.py (hunk 11 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase_requisition.py:21]: Fixed.

File: purchase_requisition_from_rfq/models/purchase_requisition.py (hunk 12 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase_requisition.py:24]: Fixed.

File: purchase_requisition_from_rfq/models/purchase_requisition.py (hunk 13 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase_requisition.py]: Fixed.

File: purchase_requisition_from_rfq/models/purchase_requisition.py (hunk 14 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase_requisition.py]: Fixed.

**Comment by ps-tubtim** [comment]
@newtratip @kittiu Thank you for your review. I fixed all. Please review again.

File: purchase_requisition_from_rfq/hooks.py (hunk 1 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/hooks.py]: Deleted.

**Review by newtratip — APPROVED** [review]
LGTM :+1:

File: purchase_requisition_from_rfq/README.rst [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+#. Select RFQs(PO's state is Draft) that need to be created the Purchase Agreement
+#. Click Action > Click Create Purchase Agreement from RFQ
+#. Confirm and Validate the purchase agreement
+#. Click smart button RFQs/Orders and Confirm order on a RFQ
+#. Return to Purchase Agreement and Click Close
+
+Note
+
+* Agreement from rfq: If you confirm order, another RFQs will cancel automatically and fill Vendor, Unit Price on Agreement follow winning Purchase Order.
+* Normal Agreement: Normal Process
```
Comment (kittiu) [inline purchase_requisition_from_rfq/README.rst]:
```suggestion
** Create Purchase Agreement from RFQs **

#. Go to menu Purchase > Orders > Requests for Quotation (RFQ)
#. Select multiple RFQ(s) that want to create purchase agreement for
#. Click Action > Create Purchase Agreement from RFQ
#. Confirm and validate the purchase agreement
#. Click smart button RFQs/Orders and confirm on the winning RFQ
#. Return to Purchase Agreement and Click Close

**Note:**

* For the agreement created from RFQ(s), when user confirm an RFQ, other RFQs will get cancelled automatically
* The agreement will be filled with, vendor and price automatically based on wining Purchase Order.

```

Comment (ps-tubtim) [reply purchase_requisition_from_rfq/README.rst]: Updated.

File: purchase_requisition_from_rfq/models/purchase.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+                    price_unit = order.order_line.filtered(
+                        lambda l: l.product_id == line.product_id
+                    ).mapped("price_unit")
+                    if price_unit:
+                        line.price_unit = max(price_unit)
+        return res
+
+    def _auto_cancel_another_order(self):
+        self.ensure_one()
+        purchases = self.env["purchase.order"].search(
```
Comment (MeetKD) [inline purchase_requisition_from_rfq/models/purchase.py]:
```suggestion
        purchases = self.search(
```

**Review by MeetKD — COMMENTED** [review]
Small Change ! Code Review :+1:

File: purchase_requisition_from_rfq/models/purchase.py (hunk 16 above)
Comment (ps-tubtim) [reply purchase_requisition_from_rfq/models/purchase.py]: @MeetKD Thank you for your review. Edited.

**Pull request closed without merging**